File: Leet 2265/solution.py
class TreeNode:
    def __init__(self, val: int = 0):
        self.val: int = val
        self.left: TreeNode|None = None
        self.right: TreeNode|None = None


class Solution:
    def get_count(self, root: TreeNode|None, ref_count: list[int]) -> tuple[int, int]:
        if not root:
            return (0, 0)
        
        left_sum, left_count = self.get_count(root.left, ref_count)
        right_sum, right_count = self.get_count(root.right, ref_count)

        subtree_sum: int = left_sum + right_sum + root.val
        subtree_count: int  = left_count + right_count + 1
        
        if root.val == (subtree_sum // subtree_count):
            ref_count[0] += 1

        return (subtree_sum, subtree_count)

    def averageOfSubtree(self, root: TreeNode) -> int:
        ref_count:list[int] = [0]
        _,_ = self.get_count(root, ref_count)
        return ref_count[0]


# -- TestCases


def case_1(sol: Solution) -> None:
    print("-- Case 1")
    root = TreeNode(4)
    root.left = TreeNode(8)
    root.right = TreeNode(5)
    root.left.left = TreeNode(0)
    root.left.right = TreeNode(1)
    root.right.right = TreeNode(6)

    sol = Solution()
    count: int = sol.averageOfSubtree(root)
    print(f"Count: {count}")


def case_2(sol: Solution) -> None:
    print("-- Case 2")
    root = TreeNode(1)
    sol = Solution()
    count: int = sol.averageOfSubtree(root)
    print(f"Count: {count}")


# -- Main


def main() -> None:
    print("2265. Count Nodes Equal to Average of Subtree")

    sol = Solution()

    case_1(sol)
    case_2(sol)


if __name__ == "__main__":
    main()

File: Leet 2265/solution.cs
Console.WriteLine("2265. Count Nodes Equal to Average of Subtree");

var root = new TreeNode(4);
root.left = new TreeNode(8);
root.right = new TreeNode(5);
root.left.left = new TreeNode(0);
root.left.right = new TreeNode(1);
root.right = new TreeNode(5);
root.right.right = new TreeNode(6);

int total = Solution.AverageOfSubtree(root);

Console.WriteLine($"Total subtree's that have an average same as root are {total}");

//---

class TreeNode
{
    public int val;
    public TreeNode? left;
    public TreeNode? right;

    public TreeNode(int val = 0)
    {
        this.val = val;
        left = null;
        right = null;
    }
}

class Solution
{
    struct Result
    { 
        public int sum; 
        public int count; 
    }

    private static Result GetResultForAverageOfSubtree(TreeNode? root, ref int total)
    {
        if (root == null)
        {
            return new Result() { sum = 0, count = 0 };
        }

        var leftResult = GetResultForAverageOfSubtree(root.left, ref total);
        var rightResult = GetResultForAverageOfSubtree(root.right, ref total);

        var sum = leftResult.sum + rightResult.sum + root.val;
        var count = leftResult.count + rightResult.count + 1;
        var average = sum / count;

        if (root.val == average)
        {
            total++;
        }

        return new Result() { sum = sum, count = count };
    }

    public static int AverageOfSubtree(TreeNode root)
    {
        int total = 0;
        _ = GetResultForAverageOfSubtree(root, ref total);
        return total;
    }
}